packages/completer-extension/src/index.ts

```diff
@@ -45,337 +41,175 @@ namespace CommandIDs {
- */
-const consoles: JupyterFrontEndPlugin<void> = {
-  id: '@jupyterlab/completer-extension:consoles',
-  requires: [ICompletionManager, IConsoleTracker],
+const defaultProvider: JupyterFrontEndPlugin<void> = {
+  id: '@jupyterlab/completer-extension:base-service',
+  requires: [ICompletionProviderManager],
   autoStart: true,
   activate: (
     app: JupyterFrontEnd,
-    manager: ICompletionManager,
-    consoles: IConsoleTracker
+    serviceManager: ICompletionProviderManager
   ): void => {
-    // Create a handler for each console that is created.
-    consoles.widgetAdded.connect((sender, widget) => {
-      const anchor = widget.console;
-      const editor = anchor.promptCell?.editor ?? null;
-      const session = anchor.sessionContext.session;
-      // TODO: CompletionConnector assumes editor and session are not null
-      const connector = new CompletionConnector({ session, editor });
-      const handler = manager.register({ connector, editor, parent: widget });
-
-      const updateConnector = () => {
-        const editor = anchor.promptCell?.editor ?? null;
-        const session = anchor.sessionContext.session;
-
-        handler.editor = editor;
-        // TODO: CompletionConnector assumes editor and session are not null
-        handler.connector = new CompletionConnector({ session, editor });
-      };
-
-      // Update the handler whenever the prompt or session changes
-      anchor.promptCellCreated.connect(updateConnector);
-      anchor.sessionContext.sessionChanged.connect(updateConnector);
-    });
-
-    // Add console completer invoke command.
-    app.commands.addCommand(CommandIDs.invokeConsole, {
-      execute: () => {
-        const id = consoles.currentWidget && consoles.currentWidget.id;
-
-        if (id) {
-          return app.commands.execute(CommandIDs.invoke, { id });
-        }
-      }
-    });
-
-    // Add console completer select command.
-    app.commands.addCommand(CommandIDs.selectConsole, {
-      execute: () => {
-        const id = consoles.currentWidget && consoles.currentWidget.id;
-
-        if (id) {
-          return app.commands.execute(CommandIDs.select, { id });
-        }
-      }
-    });
-
-    // Set enter key for console completer select command.
-    app.commands.addKeyBinding({
-      command: CommandIDs.selectConsole,
-      keys: ['Enter'],
-      selector: `.jp-ConsolePanel .jp-mod-completer-active`
-    });
+    serviceManager.registerProvider(new ContextCompleterProvider());
+    serviceManager.registerProvider(new KernelCompleterProvider());
   }
 };
 
-/**
- * An extension that registers notebooks for code completion.
- */
-const notebooks: JupyterFrontEndPlugin<void> = {
-  id: '@jupyterlab/completer-extension:notebooks',
-  requires: [ICompletionManager, INotebookTracker],
+const manager: JupyterFrontEndPlugin<ICompletionProviderManager> = {
+  id: COMPLETION_MANAGER_PLUGIN,
+  requires: [
+    INotebookTracker,
+    IEditorTracker,
+    IConsoleTracker,
```

Comment:
This gonna be a large change API wise (and the late catch). But this must be moved. JupyterLab infrastructure is based on composition of plugins. So in that spirit, this plugin should indeed provide the manager. But it is then in the notebook, file editor and console extensions that it should be used/connected to the respective editors.

This is particularly important for simplified application.

Reply:
Thanks @fcollonval, I got the point. Since I'm having a second PR in the pipeline based on this PR, can we defer this change to the second one? It would simplify the rebase work once this PR is merged.

Reply:
I applied the suggestion here https://github.com/jupyterlab/jupyterlab/pull/11969/commits/625e8ec03aef73b35f22cd75e42c0f31b93f949f